File: src/main/java/com/indiepost/repository/elasticsearch/PostEsRepository.java
package com.indiepost.repository.elasticsearch;

import com.indiepost.enums.Types;
import com.indiepost.model.User;
import com.indiepost.model.Word;
import com.indiepost.model.elasticsearch.PostEs;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface PostEsRepository {

    boolean testConnection();

    boolean indexExist();

    boolean createIndex();

    boolean deleteIndex();

    void buildIndex(List<PostEs> posts);

    void rebuildIndices(List<PostEs> posts);

    List<PostEs> search(String text, Types.PostStatus status, Pageable pageable);

    List<PostEs> search(String text, Types.PostStatus status, User currentUser, Pageable pageable);

    List<Long> moreLikeThis(List<Long> postId, Types.PostStatus status, Pageable pageable);

    List<Long> recommendation(List<Long> bookmarkedPostIds, List<Long> historyPostIds, Types.PostStatus status, Pageable pageable);

    Integer count(String text, Types.PostStatus status);

    Integer count(String text, Types.PostStatus status, User currentUser);

    PostEs findById(Long id);

    void index(PostEs post);

    void bulkIndex(List<PostEs> posts);

    void update(PostEs post);

    void deleteById(Long id);

    void delete(PostEs postEs);

    void bulkDelete(List<Long> ids);

    void updateDictionary(List<Word> words);
}

File: src/main/kotlin/com/indiepost/repository/elasticsearch/PostEsRepository.kt
package com.indiepost.repository.elasticsearch

import com.indiepost.enums.Types
import com.indiepost.model.User
import com.indiepost.model.Word
import com.indiepost.model.elasticsearch.PostEs
import org.springframework.data.domain.Pageable

interface PostEsRepository {

    fun testConnection(): Boolean

    fun indexExist(): Boolean

    fun createIndex(): Boolean

    fun deleteIndex(): Boolean

    fun buildIndex(posts: List<PostEs>)

    fun rebuildIndices(posts: List<PostEs>)

    fun search(text: String, status: Types.PostStatus, pageable: Pageable): List<PostEs>

    fun search(text: String, status: Types.PostStatus, currentUser: User, pageable: Pageable): List<PostEs>

    fun moreLikeThis(postIds: List<Long>, status: Types.PostStatus, pageable: Pageable): List<Long>

    fun recommendation(bookmarkedPostIds: List<Long>, historyPostIds: List<Long>, status: Types.PostStatus, pageable: Pageable): List<Long>

    fun count(text: String, status: Types.PostStatus): Int

    fun count(text: String, status: Types.PostStatus, currentUser: User?): Int

    fun findById(id: Long?): PostEs?

    fun index(post: PostEs)

    fun bulkIndex(posts: List<PostEs>)

    fun update(post: PostEs)

    fun deleteById(id: Long?)

    fun delete(postEs: PostEs)

    fun bulkDelete(ids: List<Long>)

    fun updateDictionary(words: List<Word>)
}
